File: Plugin/Tests/DiskDump.cpp
/*
 * The same dump as AkaiS950Tests/DiskDump.cs, from the C++ reader.
 *
 * Byte for byte the same text, so the two can be diffed. crosscheck.ps1 converts an image
 * to raw sectors with the C#, dumps it with both, and compares - which checks the port
 * against a real disk without a single image, sample name or byte of audio going into the
 * repository.
 *
 *     DiskDump <image.img>
 */

#include "Disk.h"

#include <cstdio>
#include <cstdint>
#include <string>

namespace
{
    /*
     * FNV-1a over the 16-bit words, a byte at a time, low byte first.
     *
     * A hash rather than the audio itself: the two readers can then be held to each other
     * exactly without any of someone else's recordings being written down. Unsigned
     * throughout, which is what makes it reproduce the C#'s answer.
     */
    std::string hashOf (const std::vector<short>& words)
    {
        std::uint32_t h = 2166136261u;

        for (short w : words)
        {
            const std::uint16_t u = static_cast<std::uint16_t> (w);
            h = (h ^ static_cast<std::uint32_t> (u & 0xFF)) * 16777619u;
            h = (h ^ static_cast<std::uint32_t> (u >> 8))   * 16777619u;
        }

        char buf[16];
        std::snprintf (buf, sizeof (buf), "%08x", h);
        return buf;
    }

    void printZone (const char* label, const s950::Disk::Zone& z)
    {
        std::printf ("     %s '%s' %s ptr %d fine %d trans %d filt %d loud %d\n",
                     label,
                     z.name.c_str(),
                     z.inUse() ? "used" : "unused",
                     z.pointer, z.fine, z.transpose, z.filter, z.loudness);
    }
}

int main (int argc, char** argv)
{
    if (argc < 2)
    {
        std::fprintf (stderr, "usage: DiskDump <image.img>\n");
        return 2;
    }

    s950::Disk disk;
    std::string error;

    if (! disk.loadFile (argv[1], error))
    {
        std::fprintf (stderr, "could not read %s: %s\n", argv[1], error.c_str());
        return 1;
    }

    // How the recovery went, which for an .hfe is part of what is being checked: the two
    // decoders have to agree about which sectors were bad, not just about the good ones.
    std::printf ("recovery badcrc %d missing %d\n",
                 disk.getBadCrcSectors(), disk.getMissingSectors());

    const auto& entries = disk.getEntries();
    std::printf ("entries %d\n", static_cast<int> (entries.size()));

    for (const auto& e : entries)
    {
        std::printf ("entry %d %c '%s' len %d start %d blocks %d %s\n",
                     e.slot, e.type, e.name.c_str(),
                     e.length, e.startBlock, e.chainBlocks,
                     e.chainOk ? "ok" : "SHORT");

        if (e.type == 'S')
        {
            const auto words = disk.sampleWords12 (e);

            std::printf ("  sample count %lld rate %d tuning %d loudness %d loop %c%c %lld/%lld/%lld\n",
                         static_cast<long long> (e.sampleCount), e.sampleRate, e.tuning, e.loudness,
                         e.loopMode, e.loopDirection,
                         static_cast<long long> (e.loopStart),
                         static_cast<long long> (e.loopEnd),
                         static_cast<long long> (e.loopLength));

            std::printf ("  audio words %d hash %s\n",
                         static_cast<int> (words.size()), hashOf (words).c_str());
        }
        else if (e.type == 'P')
        {
            const auto groups = disk.keygroups (e);
            std::printf ("  program keygroups %d\n", static_cast<int> (groups.size()));

            for (const auto& k : groups)
            {
                std::printf ("   kg %d keys %d-%d vsw %d vca %d,%d,%d,%d vcf %d,%d,%d,%d"
                             " amt %d vel %d,%d key %d lfo %d,%d,%d,%d flags %d\n",
                             k.index, k.lowKey, k.highKey, k.velocitySwitch,
                             k.vcaAttack, k.vcaDecay, k.vcaSustain, k.vcaRelease,
                             k.vcfAttack, k.vcfDecay, k.vcfSustain, k.vcfRelease,
                             k.vcfAmount,
                             k.velToFilter, k.velToLoudness,
                             k.keyToFilter,
                             k.lfoDelay, k.lfoRate, k.lfoDepth, k.lfoModwheelDepth,
                             k.flags);

                printZone ("z1", k.zone1);
                printZone ("z2", k.zone2);
            }
        }
    }

    return 0;
}

File: AkaiS950Tests/DiskDump.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;
using AkaiS950List;

/// <summary>
/// Print everything the reader understands about a disk, in a form another reader can be
/// held to exactly.
///
/// The C++ port in Plugin/Source/S950/Disk.cpp prints the identical text from the identical
/// bytes, and Plugin/crosscheck.ps1 diffs the two. That is how the port gets checked without
/// a single disk image, sample name or byte of audio going into the repository: the
/// reference is produced from whatever image is to hand, at the moment of checking, and
/// thrown away afterwards.
///
///     DiskDump &lt;image.hfe|image.img&gt; [--save-img &lt;path&gt;]
///
/// --save-img writes the raw 800K sectors out first. The C++ side reads only raw images so
/// far, so converting here and dumping THAT is what makes the two comparable - and it
/// exercises the conversion at the same time.
/// </summary>
static class DiskDump
{
    static int Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.Error.WriteLine("usage: DiskDump <image> [--save-img <path>]");
            return 2;
        }

        string path = args[0];
        string saveImg = null;

        for (int i = 1; i < args.Length - 1; i++)
            if (args[i] == "--save-img") saveImg = args[i + 1];

        AkaiDisk disk;
        try { disk = AkaiDisk.Load(path); }
        catch (Exception ex)
        {
            Console.Error.WriteLine("could not read " + path + ": " + ex.Message);
            return 1;
        }

        if (saveImg != null)
        {
            try { disk.SaveAs(saveImg, "img"); }
            catch (Exception ex)
            {
                Console.Error.WriteLine("could not write " + saveImg + ": " + ex.Message);
                return 1;
            }

            // Dump what the C++ will actually read, not what we started from.
            disk = AkaiDisk.Load(saveImg);
        }

        Console.Write(Dump(disk));
        return 0;
    }

    static string Dump(AkaiDisk d)
    {
        var s = new StringBuilder();

        // How the recovery went, which for an .hfe is part of what is being checked: the
        // two decoders have to agree about which sectors were bad, not just the good ones.
        s.Append("recovery badcrc ").Append(d.BadCrcSectors)
         .Append(" missing ").Append(d.MissingSectors).Append('\n');

        s.Append("entries ").Append(d.Entries.Count).Append('\n');

        foreach (AkaiEntry e in d.Entries)
        {
            s.Append("entry ")
             .Append(e.Slot).Append(' ')
             .Append(e.Type).Append(" '").Append(e.Name).Append("' ")
             .Append("len ").Append(e.Length).Append(' ')
             .Append("start ").Append(e.StartBlock).Append(' ')
             .Append("blocks ").Append(e.ChainBlocks).Append(' ')
             .Append(e.ChainOk ? "ok" : "SHORT").Append('\n');

            if (e.Type == 'S')
            {
                var words = d.SampleWords12(e);

                s.Append("  sample count ").Append(e.SampleCount)
                 .Append(" rate ").Append(e.SampleRate)
                 .Append(" tuning ").Append(e.Tuning)
                 .Append(" loudness ").Append(e.Loudness)
                 .Append(" loop ").Append(e.LoopMode).Append(e.LoopDirection)
                 .Append(' ').Append(e.LoopStart)
                 .Append('/').Append(e.LoopEnd)
                 .Append('/').Append(e.LoopLength)
                 .Append('\n');

                s.Append("  audio words ").Append(words.Length)
                 .Append(" hash ").Append(Hash(words))
                 .Append('\n');
            }
            else if (e.Type == 'P')
            {
                var groups = d.Keygroups(e);
                s.Append("  program keygroups ").Append(groups.Count).Append('\n');

                foreach (AkaiDisk.Keygroup k in groups)
                {
                    s.Append("   kg ").Append(k.Index)
                     .Append(" keys ").Append(k.LowKey).Append('-').Append(k.HighKey)
                     .Append(" vsw ").Append(k.VelocitySwitch)
                     .Append(" vca ").Append(k.VcaAttack).Append(',').Append(k.VcaDecay)
                     .Append(',').Append(k.VcaSustain).Append(',').Append(k.VcaRelease)
                     .Append(" vcf ").Append(k.VcfAttack).Append(',').Append(k.VcfDecay)
                     .Append(',').Append(k.VcfSustain).Append(',').Append(k.VcfRelease)
                     .Append(" amt ").Append(k.VcfAmount)
                     .Append(" vel ").Append(k.VelToFilter).Append(',').Append(k.VelToLoudness)
                     .Append(" key ").Append(k.KeyToFilter)
                     .Append(" lfo ").Append(k.LfoDelay).Append(',').Append(k.LfoRate)
                     .Append(',').Append(k.LfoDepth).Append(',').Append(k.LfoModwheelDepth)
                     .Append(" flags ").Append(k.Flags)
                     .Append('\n');

                    Zone(s, "z1", k.Zone1);
                    Zone(s, "z2", k.Zone2);
                }
            }
        }

        return s.ToString();
    }

    static void Zone(StringBuilder s, string label, AkaiDisk.Zone z)
    {
        s.Append("     ").Append(label);

        if (z == null)
        {
            s.Append(" -\n");
            return;
        }

        s.Append(" '").Append(z.Name).Append("' ")
         .Append(z.InUse ? "used" : "unused")
         .Append(" ptr ").Append(z.Pointer)
         .Append(" fine ").Append(z.Fine)
         .Append(" trans ").Append(z.Transpose)
         .Append(" filt ").Append(z.Filter)
         .Append(" loud ").Append(z.Loudness)
         .Append('\n');
    }

    /// <summary>
    /// FNV-1a over the 16-bit words.
    ///
    /// A hash rather than the audio itself, so the two readers can be compared exactly
    /// without any of someone else's recordings being written down. Unsigned arithmetic
    /// throughout, which is what makes it reproduce in another language.
    /// </summary>
    static string Hash(short[] words)
    {
        uint h = 2166136261;

        foreach (short w in words)
        {
            ushort u = (ushort)w;
            h = (h ^ (uint)(u & 0xFF)) * 16777619;
            h = (h ^ (uint)(u >> 8)) * 16777619;
        }

        return h.ToString("x8", CultureInfo.InvariantCulture);
    }
}
